rapid/src/main/java/com/vrg/rapid/ConsistentHashBroadcaster.java

```diff
@@ -0,0 +1,364 @@
+
+import com.google.common.util.concurrent.FutureCallback;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.SettableFuture;
+import com.google.errorprone.annotations.CanIgnoreReturnValue;
+import com.google.protobuf.ByteString;
+import com.google.protobuf.ExperimentalApi;
+import com.vrg.rapid.messaging.IBroadcaster;
+import com.vrg.rapid.messaging.IMessagingClient;
+import com.vrg.rapid.pb.BroadcastingMessage;
+import com.vrg.rapid.pb.Endpoint;
+import com.vrg.rapid.pb.FastRoundPhase2bMessage;
+import com.vrg.rapid.pb.Metadata;
+import com.vrg.rapid.pb.Proposal;
+import com.vrg.rapid.pb.RapidRequest;
+import com.vrg.rapid.pb.RapidResponse;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.GuardedBy;
+import java.util.ArrayList;
+import java.util.BitSet;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+/**
+ * Broadcaster based on a ring of broadcasting nodes organized via consistent hashing.
+ *
+ * Nodes in the cluster are segregated in two groups:
+ * - broadcasting nodes (B) that will relay messages to be broadcasted to all
+ * - "broadastee" nodes (n) that send the messages to be broadcasted to "their" broadcaster
+ *
+ * The goal is to minimize the amount of connections to other nodes that need to be maintained by a single node.
+ *
+ *
+ *                                          n n n n n
+ *                                     n        |        n
+ *                                     n        B        n
+ *                                     n ---- B   B ---- n
+ *                                     n        B        n
+ *                                     n        |        n
+ *                                          n n n n n
+ *
+ *
+ * Broadcasting works as follows:
+ * 1. A node that wants to broadcast a message sends it to it broadcaster
+ * 2. The broadcaster forwards the message to all other broadcasters and to all other of its broadcastees
+ * 3. Each broadcaster receiving a message forwards it to all of its broadcastees
+ *
+ * Additionally, fast paxos vote messages are batched together at two levels:
+ * - when first received by the broadcasting node of a group of "broadcastees"
+ * - when the broadcasted messages are subsequently received by the other broadcasters
+ *
+ * Broadcasters are discovered automatically via node meta-data. A node acts as broadcaster if configured to do so
+ * via the Settings.
+ *
+ * /!\ NOTE: This broadcasting strategy does **not preserve the order of messages**.
+ *
+ */
+@ExperimentalApi
+public class ConsistentHashBroadcaster implements IBroadcaster {
```

Comment:
The broadcaster implementation looks like a hierarchical broadcaster, where the broadcaster nodes organize themselves along a ring. I'm not sure why the ring is required given that the implementation requires broadcaster to have a special field set in their metadata (it assumes configuration)? 

The main concern however is: what happens if the number of broadcasters is less than N/4 (or N/2)? What are the resiliency properties here?

I assumed where this was going was that you'd use the 0th ring in MembershipView to deterministically organize the broadcasters. For example, deterministically pick B% out of N nodes from the 0th ring to be broadcasters and assign the remaining nodes as broadcastees (again deterministically) to a constant number of broadcasters for redundancy (so that even if a number of nodes fail, we can still make progress). Doing so has the advantage that the set of broadcasters in every membership round rotates.

Note that the expander graph has strong connectivity properties, which make it so that you can simply broadcast along it to get most of the benefits with less complexity. It also bounds the number of connections on the graph to 2K per node.

Reply:
This is a hierarchical approach indeed - it isn't the gossip-based variant from the paper.
The reason is that the broadcaster nodes are deployed on more powerful (and expensive) instances that are able to handle more concurrent connections.

I'm entirely aware of the resiliency concerns and properties of this approach.